rust/rope/src/tree.rs

```diff
@@ -116,29 +123,48 @@ enum NodeVal<N: NodeInfo> {
 
 // also consider making Metric a newtype for usize, so type system can
 // help separate metrics
+
+/// A trait for quickly processing attributes of a NodeInfo.
+/// 
+/// For the conceptual background see the 
+/// [blog post, Rope science, part 2: metrics](https://github.com/google/xi-editor/blob/master/doc/rope_science/rope_science_02.md).
 pub trait Metric<N: NodeInfo> {
+    /// Return the number of boundarys in the NodeInfo::Leaf
+    /// 
+    /// TODO: What does the second argument do?
     fn measure(&N, usize) -> usize;
 
+    /// Return a offset in base units forwich self.from_base_units will return the argument.
+    /// 
+    /// TODO: Does it have to be the largest, smallest, or just eny offset?
     fn to_base_units(l: &N::L, in_measured_units: usize) -> usize;
 
+    /// Return the number of boundarys in the NodeInfo::Leaf befor the offset in_base_units.
     fn from_base_units(l: &N::L, in_base_units: usize) -> usize;
 
-    // The next three methods work in base units.
-
-    // These methods must indicate a boundary at the end of a leaf,
-    // if present. A boundary at the beginning of a leaf is optional
-    // (the previous leaf will be queried).
-
+    /// Return whether the offset in base units is a boundary of this metric.
+    /// If a boundary is at end of a leaf then this method must return true.
+    /// However, A boundary at the beginning of a leaf is optional
+    /// (the previous leaf will be queried).
     fn is_boundary(l: &N::L, offset: usize) -> bool;
 
-    // will be called with offset > 0
+    /// Return the largest usize smaller then offset forwich is_boundary will return true.
+    /// will be called with offset > 0.
+    /// 
+    /// TODO: If offset is a boundary can this return offset, or does it have to return a strictly smaller number?
```

Comment:
"Returns the index of the boundary directly preceding `offset`, or `None` if no such boundary exists. Input and result are in base units."

Reply:
I think @cmyr has this right.